File: osudownloader.py
import os
import re
import sys
import requests

MIRRORS = [
    ("Catboy", "https://catboy.best/d/{id}"),
    ("Sayobot", "https://dl.sayobot.cn/beatmaps/download/novideo/{id}"),
    ("Nerinyan", "https://api.nerinyan.moe/d/{id}?noVideo=true"),
]

HEADERS = {"User-Agent": "Mozilla/5.0"}


def sanitize_filename(filename: str) -> str:
    clean = re.sub(r'[<>:"/\\|?*\x00-\x1f]+', '_', filename)
    clean = clean.strip('. ')
    return clean or "beatmap"


def resolve_user_id(user_input: str) -> int:
    user_input = user_input.strip()
    match = re.search(r'/(?:users|u)/(\d+)', user_input)
    if match:
        return int(match.group(1))
    if user_input.isdigit():
        return int(user_input)

    # Username lookup via profile redirect
    resp = requests.get(
        f"https://osu.ppy.sh/users/{user_input}",
        headers=HEADERS,
        allow_redirects=True,
        timeout=10,
    )
    match = re.search(r'/(?:users|u)/(\d+)', resp.url)
    if match:
        return int(match.group(1))
    raise ValueError(f"Could not resolve user ID for: '{user_input}'")


def fetch_maps(user_id: int, count: int, mode: int = 1):
    endpoint = "beatmapsets/most_played" if mode == 1 else "scores/best"
    maps = []
    offset = 0

    while len(maps) < count:
        limit = min(100, count - len(maps))
        url = f"https://osu.ppy.sh/users/{user_id}/{endpoint}?offset={offset}&limit={limit}"
        resp = requests.get(url, headers=HEADERS, timeout=15)
        if resp.status_code != 200:
            print(f"API returned HTTP {resp.status_code}")
            break
        batch = resp.json()
        if not batch:
            break
        maps.extend(batch)
        offset += len(batch)
        if len(batch) < limit:
            break

    return maps


def download_beatmapset(set_id: int, title: str, output_dir: str = "./songs") -> bool:
    safe_title = sanitize_filename(title)
    filename = f"{set_id} {safe_title}.osz"
    target_path = os.path.join(output_dir, filename)

    if os.path.exists(target_path) and os.path.getsize(target_path) > 1024:
        print(f"Already downloaded: {filename}")
        return True

    temp_path = target_path + ".tmp"
    for name, template in MIRRORS:
        url = template.format(id=set_id)
        try:
            print(f"  Downloading from {name}...", flush=True)
            with requests.get(url, headers=HEADERS, stream=True, timeout=20) as r:
                if r.status_code != 200:
                    continue
                chunks = r.iter_content(chunk_size=65536)
                first = next(chunks, b"")
                if not first.startswith(b"PK\x03\x04"):
                    continue
                with open(temp_path, "wb") as f:
                    f.write(first)
                    for chunk in chunks:
                        if chunk:
                            f.write(chunk)

            if os.path.exists(temp_path) and os.path.getsize(temp_path) > 1024:
                if os.path.exists(target_path):
                    os.remove(target_path)
                os.rename(temp_path, target_path)
                print(f"  Saved: {filename}", flush=True)
                return True
        except Exception:
            pass
        finally:
            if os.path.exists(temp_path):
                try:
                    os.remove(temp_path)
                except OSError:
                    pass

    print(f"  Failed to download mapset {set_id} from all mirrors.", flush=True)
    return False


def run_self_check():
    print("Running self-check...")
    assert sanitize_filename('test: *?/"<>| name. ') == "test_ _ name"
    assert sanitize_filename('夜に駆ける') == "夜に駆ける"
    assert resolve_user_id("https://osu.ppy.sh/users/2") == 2
    assert resolve_user_id("2") == 2
    print("Self-check passed!")


def main():
    if "--test" in sys.argv:
        run_self_check()
        return

    os.makedirs("./songs", exist_ok=True)

    if len(sys.argv) > 1:
        raw_user = sys.argv[1]
        number_of_maps = int(sys.argv[2]) if len(sys.argv) > 2 else 10
        mode = int(sys.argv[3]) if len(sys.argv) > 3 else 1
    else:
        raw_user = input("Enter User ID, Profile URL, or Username: ").strip()
        num_str = input("Enter number of maps to download (default 10): ").strip()
        number_of_maps = int(num_str) if num_str else 10
        mode_str = input("Choose mode [1: Most Played (default), 2: Top Plays (Best pp)]: ").strip()
        mode = int(mode_str) if mode_str in ("1", "2") else 1

    try:
        user_id = resolve_user_id(raw_user)
    except Exception as e:
        print(f"Error: {e}")
        return

    mode_label = "Most Played" if mode == 1 else "Top Plays (Best pp)"
    print(f"\nFetching {number_of_maps} {mode_label} for user ID {user_id}...")
    raw_maps = fetch_maps(user_id, number_of_maps, mode)
    if not raw_maps:
        print("No beatmaps found.")
        return

    # Deduplicate beatmapsets
    seen = set()
    mapsets = []
    for item in raw_maps:
        set_info = item.get("beatmapset")
        if set_info and set_info["id"] not in seen:
            seen.add(set_info["id"])
            mapsets.append((set_info["id"], set_info["title"]))

    print(f"Found {len(mapsets)} unique beatmapsets to download.\n")

    success = 0
    for idx, (sid, title) in enumerate(mapsets, 1):
        print(f"[{idx}/{len(mapsets)}] Beatmapset {sid}: {title}")
        if download_beatmapset(sid, title):
            success += 1

    print(f"\nDone! Downloaded {success}/{len(mapsets)} beatmapsets into './songs/'.")


if __name__ == "__main__":
    main()

File: OSUDownloader.js
const fs = require('node:fs');
const path = require('node:path');
const { pipeline } = require('node:stream/promises');
const { Readable } = require('node:stream');
const readline = require('node:readline/promises');

const MIRRORS = [
  { name: 'Catboy', url: (id) => `https://catboy.best/d/${id}` },
  { name: 'Sayobot', url: (id) => `https://dl.sayobot.cn/beatmaps/download/novideo/${id}` },
  { name: 'Nerinyan', url: (id) => `https://api.nerinyan.moe/d/${id}?noVideo=true` },
];

const HEADERS = { 'User-Agent': 'Mozilla/5.0' };

function sanitizeFilename(filename) {
  const clean = filename.replace(/[<>:"/\\|?*\x00-\x1f]+/g, '_').replace(/^[. ]+|[. ]+$/g, '');
  return clean || 'beatmap';
}

async function resolveUserId(rawUser) {
  const trimmed = rawUser.trim();
  const urlMatch = trimmed.match(/\/(?:users|u)\/(\d+)/);
  if (urlMatch) return parseInt(urlMatch[1], 10);
  if (/^\d+$/.test(trimmed)) return parseInt(trimmed, 10);

  // Username lookup via profile redirect
  const res = await fetch(`https://osu.ppy.sh/users/${encodeURIComponent(trimmed)}`, {
    headers: HEADERS,
    redirect: 'follow',
  });
  const match = res.url.match(/\/(?:users|u)\/(\d+)/);
  if (match) return parseInt(match[1], 10);
  throw new Error(`Could not resolve user ID for: '${trimmed}'`);
}

async function fetchMaps(userId, count, mode = 1) {
  const endpoint = mode === 1 ? 'beatmapsets/most_played' : 'scores/best';
  const maps = [];
  let offset = 0;

  while (maps.length < count) {
    const limit = Math.min(100, count - maps.length);
    const url = `https://osu.ppy.sh/users/${userId}/${endpoint}?offset=${offset}&limit=${limit}`;
    const res = await fetch(url, { headers: HEADERS });
    if (!res.ok) {
      console.log(`API returned HTTP ${res.status}`);
      break;
    }
    const batch = await res.json();
    if (!Array.isArray(batch) || batch.length === 0) break;
    maps.push(...batch);
    offset += batch.length;
    if (batch.length < limit) break;
  }

  return maps;
}

function isValidZip(filePath) {
  try {
    const fd = fs.openSync(filePath, 'r');
    const buf = Buffer.alloc(4);
    fs.readSync(fd, buf, 0, 4, 0);
    fs.closeSync(fd);
    return buf[0] === 0x50 && buf[1] === 0x4b && buf[2] === 0x03 && buf[3] === 0x04;
  } catch {
    return false;
  }
}

async function downloadBeatmapset(setId, title, outputDir = './songs') {
  const safeTitle = sanitizeFilename(title);
  const filename = `${setId} ${safeTitle}.osz`;
  const targetPath = path.join(outputDir, filename);

  if (fs.existsSync(targetPath) && fs.statSync(targetPath).size > 1024) {
    console.log(`Already downloaded: ${filename}`);
    return true;
  }

  const tempPath = targetPath + '.tmp';
  for (const mirror of MIRRORS) {
    const url = mirror.url(setId);
    try {
      console.log(`  Downloading from ${mirror.name}...`);
      const res = await fetch(url, { headers: HEADERS });
      if (!res.ok || !res.body) continue;

      await pipeline(Readable.fromWeb(res.body), fs.createWriteStream(tempPath));

      if (fs.existsSync(tempPath) && fs.statSync(tempPath).size > 1024 && isValidZip(tempPath)) {
        if (fs.existsSync(targetPath)) {
          fs.unlinkSync(targetPath);
        }
        fs.renameSync(tempPath, targetPath);
        console.log(`  Saved: ${filename}`);
        return true;
      }
    } catch {
      // try next mirror
    } finally {
      if (fs.existsSync(tempPath)) {
        try {
          fs.unlinkSync(tempPath);
        } catch {}
      }
    }
  }

  console.log(`  Failed to download mapset ${setId} from all mirrors.`);
  return false;
}

async function runSelfCheck() {
  console.log('Running self-check...');
  if (sanitizeFilename('test: *?/"<>| name. ') !== 'test_ _ name') throw new Error('sanitizeFilename failed');
  if (sanitizeFilename('夜に駆ける') !== '夜に駆ける') throw new Error('unicode sanitizeFilename failed');
  const uid = await resolveUserId('2');
  if (uid !== 2) throw new Error('resolveUserId failed');
  console.log('Self-check passed!');
}

async function main() {
  if (process.argv.includes('--test')) {
    await runSelfCheck();
    return;
  }

  fs.mkdirSync('./songs', { recursive: true });

  let rawUser = '';
  let numberOfMaps = 10;
  let mode = 1;

  if (process.argv.length > 2) {
    rawUser = process.argv[2];
    numberOfMaps = process.argv[3] ? parseInt(process.argv[3], 10) : 10;
    mode = process.argv[4] ? parseInt(process.argv[4], 10) : 1;
  } else {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    rawUser = (await rl.question('Enter User ID, Profile URL, or Username: ')).trim();
    const numStr = (await rl.question('Enter number of maps to download (default 10): ')).trim();
    numberOfMaps = numStr ? parseInt(numStr, 10) : 10;
    const modeStr = (await rl.question('Choose mode [1: Most Played (default), 2: Top Plays (Best pp)]: ')).trim();
    mode = modeStr === '2' ? 2 : 1;

    rl.close();
  }

  let userId;
  try {
    userId = await resolveUserId(rawUser);
  } catch (err) {
    console.error(`Error: ${err.message}`);
    return;
  }

  const modeLabel = mode === 1 ? 'Most Played' : 'Top Plays (Best pp)';
  console.log(`\nFetching ${numberOfMaps} ${modeLabel} for user ID ${userId}...`);
  const rawMaps = await fetchMaps(userId, numberOfMaps, mode);

  if (!rawMaps || rawMaps.length === 0) {
    console.log('No beatmaps found.');
    return;
  }

  // Deduplicate beatmapsets
  const seen = new Set();
  const mapsets = [];
  for (const item of rawMaps) {
    const setInfo = item.beatmapset;
    if (setInfo && !seen.has(setInfo.id)) {
      seen.add(setInfo.id);
      mapsets.push({ id: setInfo.id, title: setInfo.title });
    }
  }

  const totalDifficulties = rawMaps.length;
  console.log(`Found ${mapsets.length} unique beatmapsets (${totalDifficulties} total difficulties/plays) to download.\n`);

  let success = 0;
  for (let idx = 0; idx < mapsets.length; idx++) {
    const { id: sid, title } = mapsets[idx];
    console.log(`[${idx + 1}/${mapsets.length}] Beatmapset ${sid}: ${title}`);
    if (await downloadBeatmapset(sid, title)) {
      success++;
    }
  }

  console.log(`\nDone! Downloaded ${success}/${mapsets.length} (${totalDifficulties} difficulties) beatmapsets into './songs/'.`);
}

main().catch(console.error);
